### PR TITLE
Add a message method to print a message above the progress bar

### PR DESCRIPTION
**Pull request by jimhester**
https://asciinema.org/a/KORGKhrqmPBdAKbAH3u8iUN76

### CODE REVIEW

File: R/progress.R [hunk 1]
```diff
@@ -397,7 +400,17 @@ pb_update <- function(self, private, ratio, tokens) {
   self$tick(goal - private$current, tokens)
 }
 
+pb_message <- function(self, private, msg) {
+  clear_line(private$stream, private$width)
+  cursor_to_start(private$stream)
+  cat(msg, sep = "\n", file = private$stream)
+  if (!private$finished) {
```
Comment (gaborcsardi) [inline R/progress.R]: We also need `&& private$supported` I think.

**Comment by gaborcsardi** [comment]
Actually, do we want to omit the messages as well if `!supported`?

**Comment by gaborcsardi** [comment]
ACtually, if the message is coming after we have finished, then we might not need to clear the line, or do we?

**Comment by jimhester** [comment]
Yeah you are right, I guess we need to decide if it makes sense to have a message after we are done or not.

Also right now if you call `tick()` on a completed bar, it flashes for a second then goes away, maybe this should just throw an error?

**Comment by jimhester** [comment]
Not sure what we should do if a progress bar is not supported, I guess messages might still be OK in that case...

**Comment by gaborcsardi** [comment]
I think we could argue that the program needs to do all the messages before the bar is completed. And then additional messages and ticks will error. 

Because if you can message after done, it is hard to control the screen.

**Comment by gaborcsardi** [comment]
Yeah, I think we can keep the messages if `!supported`.

**Comment by jimhester** [comment]
So another thing that might be useful is to use tokens in messages, like `pb$message(finished xyz :elapsed)`, but it wasn't immediately obvious to me what tokens are valid for a given progress bar or if you should be able to have tokens in messages that aren't in the progress bar etc.

**Comment by gaborcsardi** [comment]
Let's just leave the tokens out for now. It is not too hard to add them, but would need some refactoring if we want to allow arbitrary tokens there. Currently I only update the tokens that are included in the progress bar.

**Comment by jimhester** [comment]
I added simple `stopifnot()` tests. I can use `stop()` with a better error message, just wasn't sure what style you were using for this package, all the other errors are using stopifnot assertions.

File: R/progress.R [hunk 2]
```diff
@@ -393,11 +397,26 @@ pb_render <- function(self, private, tokens) {
...
 }
 
+pb_message <- function(self, private, msg) {
+  assert_character(msg)
+  stopifnot(!private$finished)
+
+  clear_line(private$stream, private$width)
+  cursor_to_start(private$stream)
+  cat(msg, sep = "\n", file = private$stream)
+  if (!private$finished) {
```
Comment (gaborcsardi) [inline R/progress.R]: Can you pls add `&& private$supported` here?

Comment (gaborcsardi) [reply R/progress.R]: Actually, I think we can do a big `if` here. We might not be able to clear the line is `!supported`.

Comment (gaborcsardi) [reply R/progress.R]: So the current code is fine if `supported`, otherwise just `cat` the message and a newline, I think.

Comment (jimhester) [reply R/progress.R]: Done, sorry!

**Comment by jimhester** [comment]
Sorry, ok I should really be fixed this time...

File: R/progress.R [hunk 3]
```diff
@@ -393,11 +397,30 @@ pb_render <- function(self, private, tokens) {
...
+  assert_character(msg)
+  stopifnot(!private$finished)
+
+  if (!private$supported) {
+    cat(msg, sep = "\n", file = private$stream)
+  } else {
+    clear_line(private$stream, private$width)
+    cursor_to_start(private$stream)
+    cat(msg, sep = "\n", file = private$stream)
+    if (!private$finished) {
```
Comment (gaborcsardi) [inline R/progress.R]: You don't need to check for this, because you checked it up front already.

**Pull request merged**